### PR TITLE
Add writeable_on flag for attributes

### PR DESCRIPTION
**Pull request by lcpriest**

### CODE REVIEW

**Comment by lcpriest** [comment]
Thoughts @bradleypriest

File: lib/gecko/helpers/serialization_helper.rb [hunk 1]
```diff
@@ -28,11 +28,30 @@ def serializable_hash
...
+          next unless writeable?(attribute)
           serialize_attribute(attribute_hash, attribute)
         end
         attribute_hash
       end
 
+      # Returns true if an attribute can be serialized
+      #
+      # @example
+      #   product.readonly(:initial_cost_price) #=> true
```
Comment (bradleypriest) [inline lib/gecko/helpers/serialization_helper.rb]: This example is incorrect?

File: lib/gecko/helpers/serialization_helper.rb [hunk 2]
```diff
@@ -28,11 +28,31 @@ def serializable_hash
...
+          next unless writeable?(attribute)
           serialize_attribute(attribute_hash, attribute)
         end
         attribute_hash
       end
 
+      # Returns true if an attribute can be serialized
+      #
+      # @example
+      #   attribute = Gecko::Record::Base.attribute :initial_cost_price, writeable_on: :create
```
Comment (bradleypriest) [inline lib/gecko/helpers/serialization_helper.rb]: You're right it is funky, It's private, I'm not sure if it needs an example

File: lib/gecko/helpers/serialization_helper.rb [hunk 3]
```diff
@@ -28,11 +28,27 @@ def serializable_hash
...
         end
         attribute_hash
       end
 
+      # Returns true if an attribute can be serialized
+      #
+      # @return [Boolean]
+      #
+      # @api private
+      def writeable?(attribute)
```
Comment (bradleypriest) [inline lib/gecko/helpers/serialization_helper.rb]: Should we move the readonly check in here as well?

**Comment by bradleypriest** [comment]
The API for this still seems wrong to me

**Comment by lcpriest** [comment]
@wassimk do you have any thoughts on this? 

I don't like the name of the API, but I want the gem to handle this without having to be rejected by the API. The idea is that there are some attributes you can set on create, but you are not allowed to update later on.

**Comment by wassimk** [comment]
Hey @lcpriest, I like it! Would help to get rejected in the GEM for sure. I really don't like getting those API rejections via oauth objects.

**Comment by lcpriest** [comment]
At the moment we don't like the naming, have you worked with other gems with something similar/do you have any thoughts on what better naming could be?
`writeable_on` is not something I've seen before.

**Comment by wassimk** [comment]
I actually sat here for about five minutes on my last reply trying to think of something better. It doesn't sound right but I can't really come up with a better name. `editable` is the closet I could think of but I'm not sure it's any better. One possibility is rework of it like Rails's callbacks?

```ruby
after_initialize :set_stripe_customer_id, if: :new_record?
after_commit :update_order_in_tradegecko, on: [:create, :update]
```

**Comment by lcpriest** [comment]
I tried the `on: :create` syntax earlier. I like it, but I don't think it is as clear as `writeable_on:`.

`belongs_to :order, on: :create` isn't really the same as 
`after_commit :commit_something, on: :create` as the belongs_to isn't a life cycle hook, so you aren't expecting it to *do* any action.

**Comment by wassimk** [comment]
I agree. Sometimes there is nothing better. When I first saw it I knew what to expect before even looking at the implementation. From my perspective that's a win.

File: lib/gecko/helpers/serialization_helper.rb (hunk 3 above)
Comment (lcpriest) [inline lib/gecko/helpers/serialization_helper.rb]: Yup, good idea.

File: lib/gecko/helpers/serialization_helper.rb [hunk 4]
```diff
@@ -27,12 +27,28 @@ def as_json
...
+        return false if attribute.options[:readonly]
+        return true unless attribute.options[:writeable_on]
```
Comment (bradleypriest) [inline lib/gecko/helpers/serialization_helper.rb]: Can we just add this as a third when to the case, this shouldn't be a performance bottleneck, so we can go with explicit

File: lib/gecko/helpers/serialization_helper.rb [hunk 5]
```diff
@@ -27,12 +27,29 @@ def as_json
...
+      #
+      # @api private
+      def writeable?(attribute)
+        return false if attribute.options[:readonly]
+        case attribute.options[:writeable_on]
+        when :update
+          persisted?
+        when :create
+          !persisted?
+        else
```
Comment (bradleypriest) [inline lib/gecko/helpers/serialization_helper.rb]: Maybe raise on `else`

Comment (lcpriest) [inline lib/gecko/helpers/serialization_helper.rb]: Done

**Comment by MohammadHabbab** [comment]
LGTM :+1:

**Pull request merged**